src/operator/setup.rs

```diff
@@ -0,0 +1,185 @@
+pub struct TestClusterTimeouts {
+    pub cluster_ready: Duration,
+    pub pods_terminated: Duration,
+}
+
+impl<T> TestCluster<T>
+where
+    T: Clone + Debug + DeserializeOwned + Resource<DynamicType = ()> + Serialize,
+{
+    /// This creates a kube client and should be executed at the start of every test.
+    pub fn new(options: TestClusterOptions, timeouts: TestClusterTimeouts) -> Self {
+        TestCluster {
+            client: TestKubeClient::new(),
+            cluster: None,
+            options,
+            timeouts,
+        }
+    }
+
+    /// Applies a custom resource, stores the returned cluster object and sleeps for
+    /// to seconds to give the operator time to react on the custom resource.
+    /// Without the sleep it can happen that tests run without any pods being created.
+    fn apply(&mut self, cluster: &T) -> Result<()> {
+        self.cluster = Some(self.client.apply(&serde_yaml::to_string(cluster)?));
+
+        // we wait here to give the operator time to react to the custom resource
+        thread::sleep(Duration::from_secs(2));
+        Ok(())
+    }
+
+    /// Applies a command and waits 2 seconds to let the operator react on in.
+    pub fn apply_command<C>(&self, command: &C) -> Result<C>
+    where
+        C: Clone + Debug + DeserializeOwned + Resource<DynamicType = ()> + Serialize,
+    {
+        let cmd: C = self.client.apply(&serde_yaml::to_string(command)?);
+
+        // we wait here to give the operator time to react to the command
+        thread::sleep(Duration::from_secs(2));
+        Ok(cmd)
+    }
+
+    /// Deletes the custom resource, and waits for pods to be terminated
+    /// This should be executed after every single test to provide a "clean"
+    /// environment for the following tests.
+    pub fn delete(&mut self) -> Result<()> {
+        if let Some(cluster) = self.cluster.take() {
+            self.client.delete(cluster);
+            self.wait_for_pods_terminated()?;
+            self.cluster = None;
+        }
+
+        Ok(())
+    }
+
+    /// Delete a command resource. This is for clean up purposes.
+    pub fn delete_command<C>(&mut self, command: C) -> Result<()>
+    where
+        C: Clone + Debug + DeserializeOwned + Resource<DynamicType = ()> + Serialize,
+    {
+        self.client.delete(command);
+        Ok(())
+    }
+
+    /// Returns all available pods in the cluster via the name label.
+    pub fn get_current_pods(&self) -> Vec<Pod> {
+        let current_pods: ObjectList<Pod> = self.client.list_labeled(&format!(
+            "app.kubernetes.io/name={}",
+            &self.options.pod_name_label
+        ));
+        current_pods.items
+    }
+
+    /// Returns all available nodes in the cluster. Can be used to determine the expected pods
+    /// for tests (depending on the custom resource)
+    pub fn get_available_nodes(&self) -> Vec<Node> {
+        let available_nodes: ObjectList<Node> = self.client.list_labeled("");
```

Comment:
If only Stackable nodes should be returned then objects labeled with `kubernetes.io/arch=stackable-linux` should be listed.